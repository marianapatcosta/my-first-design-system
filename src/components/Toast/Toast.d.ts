import React, { CSSProperties } from 'react';
import { TOAST_TYPE } from '../../constants';
declare type ToastType = TOAST_TYPE;
export interface ToastProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
    /**
     * Toast message
     */
    message: string;
    /**
     * Toast type; there are 4 types available: alert, info, success and warning
     */
    type?: ToastType;
    /**
     * Time to toast autodelete, in milliseconds
     */
    autoClearTime?: number;
    /**
     * Handles event on Toast autodelete
     */
    onClear: () => void;
}
export declare const Toast: React.FC<ToastProps>;
export {};
