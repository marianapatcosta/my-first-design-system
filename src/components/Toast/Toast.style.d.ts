import { MouseEventHandler } from 'react';
export declare const StyledToast: import("styled-components").StyledComponent<"div", any, {
    type?: string | undefined;
    willBeDeleted?: boolean | undefined;
}, never>;
export declare const StyledToastClose: import("styled-components").StyledComponent<import("react").ForwardRefExoticComponent<import("../Button/Button").ButtonProps & import("react").RefAttributes<HTMLButtonElement>>, any, {
    icon?: string | undefined;
    onClick?: MouseEventHandler<HTMLButtonElement> | undefined;
}, never>;
export declare const StyledToastIcon: import("styled-components").StyledComponent<"img", any, {}, never>;
export declare const StyledToastMessage: import("styled-components").StyledComponent<"div", any, {
    type?: string | undefined;
}, never>;
export declare const StyledToastMessageWrapper: import("styled-components").StyledComponent<"div", any, {}, never>;
