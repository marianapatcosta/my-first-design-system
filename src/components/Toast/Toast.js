var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useState, } from 'react';
import { Alert, CloseWhite, Info, Success, Warning } from '../../assets/icons';
import { TOAST_TYPE } from '../../constants';
import { StyledToast, StyledToastClose, StyledToastIcon, StyledToastMessage, StyledToastMessageWrapper, } from './Toast.style';
var toastIcons = {
    alert: Alert,
    info: Info,
    success: Success,
    warning: Warning,
};
export var Toast = function (_a) {
    var message = _a.message, type = _a.type, className = _a.className, autoClearTime = _a.autoClearTime, onClear = _a.onClear, props = __rest(_a, ["message", "type", "className", "autoClearTime", "onClear"]);
    var _b = useState(false), willBeDeleted = _b[0], setWillBeDeleted = _b[1];
    useEffect(function () {
        var autoDeleteTimer;
        var fadeOutTimer;
        if (message) {
            var fadeOutTime = autoClearTime && autoClearTime - 500;
            fadeOutTimer = setTimeout(function () { return setWillBeDeleted(true); }, fadeOutTime);
            autoDeleteTimer = setTimeout(function () { return onClear && onClear(); }, autoClearTime);
        }
        return function () {
            clearTimeout(fadeOutTimer);
            clearTimeout(autoDeleteTimer);
        };
    }, [autoClearTime, message, onClear]);
    onClear;
    return (_jsxs(StyledToast, __assign({ className: className, type: type }, props, { willBeDeleted: willBeDeleted }, { children: [_jsx(StyledToastClose, { icon: CloseWhite, "aria-label": 'close toast', size: 'small', onClick: onClear }, void 0),
            _jsxs(StyledToastMessageWrapper, { children: [_jsx(StyledToastIcon, { src: type && toastIcons[type], alt: (type && toastIcons[type]) + " icon" }, void 0),
                    _jsx(StyledToastMessage, __assign({ type: type }, { children: message }), void 0)] }, void 0)] }), void 0));
};
Toast.defaultProps = {
    id: '',
    className: '',
    style: {},
    message: '',
    type: TOAST_TYPE.INFO,
    autoClearTime: 2500,
    onClear: function () { return null; },
};
