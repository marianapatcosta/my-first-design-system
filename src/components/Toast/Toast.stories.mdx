import { Meta, Story, ArgsTable, Canvas, Preview } from '@storybook/addon-docs/blocks';
import { Toast } from "./Toast";

<Meta title="Design System/Toast" component={Toast} argTypes={{
    type: {
      control: {
        type: 'radio',
        options: ['info', 'alert', 'warning', 'success']
      }
    }
  }}/>

# Toast
<br/><br/> 

### Displays a temporary message on the screen

Use Toast to display a temporary message for the user. There are 4 toast types available:
_alert_, _info_, _success_ and _warning_

<Preview withToolbar>
  <Story name="Standard">
    <Toast message="info toast" style={{position: 'relative', top: 'revert'}}/>
  </Story>
</Preview>

<ArgsTable of={Toast} />

### Alert Toast:

  <Story name="Alert">
    <Toast message="alert toast" type="alert" style={{position: 'relative', top: 'revert', marginLeft: '1rem'}}/>
  </Story>

### Warning Toast:

  <Story name="Warning">
    <Toast message="warning toast" type="warning" style={{position: 'relative', top: 'revert', marginLeft: '1rem'}}/>
  </Story>

### Success Toast:
  <Story name="Success">
    <Toast message="success toast" type="success" style={{position: 'relative', top: 'revert', marginLeft: '1rem'}} />
  </Story>

### Playground

Experiment with Toast Component with the Controls addon in the Canvas tab.

export const Template = (args) => <Toast {...args} />

<Canvas>
  <Story name="Controls" args={{
   message: 'toast message',
   style: {position: 'relative', top: 'revert', marginLeft: '1rem'}
  }}>
    {Template.bind({})}
  </Story>
</Canvas> 