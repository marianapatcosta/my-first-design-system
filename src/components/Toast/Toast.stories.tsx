import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Toast, ToastProps } from './Toast';
import { TOAST_TYPE } from '../../constants';

export default {
  title: 'Toast',
  component: Toast,
  args: {},
} as Meta;

const Template: Story<ToastProps> = (args) => <Toast {...args} />;

export const Info = Template.bind({});
Info.args = { message: 'info toast', type: TOAST_TYPE.INFO };

export const Alert = Template.bind({});
Alert.args = { message: 'alert toast', type: TOAST_TYPE.ALERT };

export const Warning = Template.bind({});
Warning.args = { message: 'warning toast', type: TOAST_TYPE.WARNING };

export const Success = Template.bind({});
Success.args = { message: 'success toast', type: TOAST_TYPE.SUCCESS };
