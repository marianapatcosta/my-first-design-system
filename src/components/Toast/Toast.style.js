var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled, { css, keyframes } from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { Button } from '../Button/Button';
var typeClass = function (type) {
    var toastTypesClasses = {
        alert: colors.red,
        info: colors.blue,
        success: colors.green,
        warning: colors.yellow,
    };
    return toastTypesClasses[type];
};
var fadeIn = keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  0% {\n    opacity: 0;\n  }\n  100% {\n    opacity: 1;\n  }\n"], ["\n  0% {\n    opacity: 0;\n  }\n  100% {\n    opacity: 1;\n  }\n"])));
var fadeOut = keyframes(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  0% {\n    opacity: 1;\n  }\n  100% {\n    opacity: 0;\n  }\n"], ["\n  0% {\n    opacity: 1;\n  }\n  100% {\n    opacity: 0;\n  }\n"])));
export var StyledToast = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n ", "\n  box-shadow: 0 0.0625rem 0.2rem ", ";\n  width: 85%;\n  max-width: 20rem;\n  position: fixed;\n  top: 4rem;\n  right: 1rem;\n  padding: 1rem;\n  z-index: 30;\n  display: flex;\n  align-items: flex-start;\n  border-radius: 0.2rem;\n  animation: ", " 0.5s;\n  -webkit-animation: ", " 0.5s;\n  background-color: ", ";\n\n  ", "\n\n  @media (min-width: 480px) {\n    width: 75%;\n  }\n\n  @media (min-width: 768px) {\n   width: 50%;\n  }\n\n  @media (min-width: 1023px) {\n   width: 30%;\n  }\n"], ["\n ", "\n  box-shadow: 0 0.0625rem 0.2rem ", ";\n  width: 85%;\n  max-width: 20rem;\n  position: fixed;\n  top: 4rem;\n  right: 1rem;\n  padding: 1rem;\n  z-index: 30;\n  display: flex;\n  align-items: flex-start;\n  border-radius: 0.2rem;\n  animation: ", " 0.5s;\n  -webkit-animation: ", " 0.5s;\n  background-color: ", ";\n\n  ",
    "\n\n  @media (min-width: 480px) {\n    width: 75%;\n  }\n\n  @media (min-width: 768px) {\n   width: 50%;\n  }\n\n  @media (min-width: 1023px) {\n   width: 30%;\n  }\n"])), baseStyles, colors.disabled, fadeIn, fadeIn, function (_a) {
    var type = _a.type;
    return !!type && typeClass(type);
}, function (_a) {
    var willBeDeleted = _a.willBeDeleted;
    return willBeDeleted && css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n      animation: ", " 0.5s;\n      -webkit-animation: ", " 0.5s;\n    "], ["\n      animation: ", " 0.5s;\n      -webkit-animation: ", " 0.5s;\n    "])), fadeOut, fadeOut);
});
export var StyledToastClose = styled(Button)(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  position: absolute;\n  top: 0.5rem;\n  right: 0.5rem;\n  border: none;\n  img {\n    width: 0.7rem;\n    transform: rotate(0deg);\n    transition: transform 0.3s ease-in-out;\n\n    &:hover {\n      transform: rotate(90deg);\n    }\n  }\n"], ["\n  position: absolute;\n  top: 0.5rem;\n  right: 0.5rem;\n  border: none;\n  img {\n    width: 0.7rem;\n    transform: rotate(0deg);\n    transition: transform 0.3s ease-in-out;\n\n    &:hover {\n      transform: rotate(90deg);\n    }\n  }\n"])));
export var StyledToastIcon = styled.img(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  width: 1.5rem;\n  height: 1.5rem;\n"], ["\n  width: 1.5rem;\n  height: 1.5rem;\n"])));
export var StyledToastMessage = styled.div(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  font-size: 95%;\n  color: ", ";\n  line-height: 1.5rem;\n  padding: 0 1rem;\n"], ["\n  font-size: 95%;\n  color: ", ";\n  line-height: 1.5rem;\n  padding: 0 1rem;\n"])), colors.white);
export var StyledToastMessageWrapper = styled.div(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  margin-top: 0.5rem;\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  margin-top: 0.5rem;\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8;
