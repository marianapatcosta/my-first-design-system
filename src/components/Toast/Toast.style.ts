import { MouseEventHandler } from 'react';
import styled, { css, keyframes } from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { Button } from '../Button/Button';

const typeClass = (type: string): string => {
  const toastTypesClasses: any = {
    alert: colors.red,
    info: colors.blue,
    success: colors.green,
    warning: colors.yellow,
  };
  return toastTypesClasses[type];
};

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

const fadeOut = keyframes`
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
`;

export const StyledToast = styled.div<{
  type?: string;
  willBeDeleted?: boolean;
}>`
 ${baseStyles}
  box-shadow: 0 0.125rem 0.5rem ${colors.shadow};
  width: 85%;
  max-width: 20rem;
  position: absolute;
  top: 35%;
  left: 0;
  right: 0;
  margin: 0 auto;
  padding: 1rem;
  z-index: 2;
  display: flex;
  align-items: flex-start;
  border-radius: 0.2rem;
  animation: ${fadeIn} 0.5s;
  -webkit-animation: ${fadeIn} 0.5s;
  background-color: ${({ type }) => !!type && typeClass(type)};

  ${({ willBeDeleted }) =>
    willBeDeleted &&
    css`
      animation: ${fadeOut} 0.5s;
      -webkit-animation: ${fadeOut} 0.5s;
    `}

  @media (min-width: 480px) {
    width: 75%;
  }

  @media (min-width: 768px) {
   width: 50%;
  }

  @media (min-width: 1023px) {
   width: 30%;
  }
`;

export const StyledToastClose = styled(Button)<{
  icon?: string;
  onClick?: MouseEventHandler<HTMLButtonElement>;
}>`
  position: absolute;
  top: 0;
  right: 0;
  border: none;
  img {
    ${colors.icon};
    transform: rotate(0deg);
    transition: transform 0.3s ease-in-out;

    &:hover {
      transform: rotate(90deg);
    }
  }
`;

export const StyledToastIcon = styled.img`
  width: 1.7rem;
  height: 1.7rem;
`;

export const StyledToastMessage = styled.div<{ type?: string }>`
  font-size: 95%;
  color: ${colors.white};
  font-weight: 700;
  padding: 0 1rem;
`;
