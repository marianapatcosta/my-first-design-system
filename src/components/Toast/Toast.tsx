import React, {
  CSSProperties,
  MouseEventHandler,
  useEffect,
  useState,
} from 'react';
import { Alert, Close, Info, Success, Warning } from '../../assets/icons';
import {
  StyledToast,
  StyledToastClose,
  StyledToastIcon,
  StyledToastMessage,
} from './Toast.style';

const toastIcons: {
  alert: string;
  info: string;
  success: string;
  warning: string;
} = {
  alert: Alert,
  info: Info,
  success: Success,
  warning: Warning,
};

type ToastType = 'alert' | 'info' | 'success' | 'warning';

export interface ToastProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
  /**
   * Toast message
   */
  message: string;
  /**
   * Toast type; there are 4 types available: alert, info, success and warning
   */
  type?: ToastType;
  /**
   * Time to toast autodelete, in milliseconds
   */
  autoClearTime?: number;
  /**
   * Handles event on Toast autodelete
   */
  onClear: () => void;
}

export const Toast: React.FC<ToastProps> = ({
  message,
  type,
  className,
  autoClearTime,
  onClear,
  ...props
}) => {
  const [willBeDeleted, setWillBeDeleted] = useState<boolean>(false);

  useEffect(() => {
    let autoDeleteTimer: number;
    let fadeOutTimer: number;
    if (message) {
      const fadeOutTime = autoClearTime && autoClearTime - 500;
      fadeOutTimer = setTimeout(() => setWillBeDeleted(true), fadeOutTime);
      autoDeleteTimer = setTimeout(() => onClear && onClear(), autoClearTime);
    }
    return () => {
      clearTimeout(fadeOutTimer);
      clearTimeout(autoDeleteTimer);
    };
  }, [autoClearTime, message, onClear]);

  onClear as MouseEventHandler<HTMLButtonElement>;
  return (
    <StyledToast
      className={className}
      type={type}
      {...props}
      willBeDeleted={willBeDeleted}
    >
      <StyledToastClose icon={Close} size='small' onClick={onClear} />
      <StyledToastIcon
        src={type && toastIcons[type]}
        alt={`${type && toastIcons[type]} icon`}
      />
      <StyledToastMessage type={type}>{message}</StyledToastMessage>
    </StyledToast>
  );
};

Toast.defaultProps = {
  id: '',
  className: '',
  style: {},
  message: '',
  type: 'info',
  autoClearTime: 2500,
  onClear: () => null,
};
