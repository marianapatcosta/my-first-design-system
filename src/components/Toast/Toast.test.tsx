import React from 'react';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { Toast } from './Toast';

describe('Toast', () => {
  const defaultProps = { message: 'info toast', onClear: () => null };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Toast {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display message', () => {
    const { getByText } = render(<Toast {...defaultProps} />);

    getByText(/Toast/i);
  });

  test('should handle onClear events when message is truthy', () => {
    jest.useFakeTimers();
    const onClearMock = jest.fn();
    const { container } = render(
      <Toast {...defaultProps} onClear={onClearMock} />
    );
    act(() => {
      jest.runAllTimers();
    });
    jest.useRealTimers();
    expect(onClearMock).toBeCalled();
  });

  test('should not handle onClear events when message is falsy', () => {
    jest.useFakeTimers();
    const onClearMock = jest.fn();
    const { container } = render(
      <Toast {...defaultProps} message='' onClear={onClearMock} />
    );
    act(() => {
      jest.runAllTimers();
    });
    jest.useRealTimers();
    expect(onClearMock).not.toBeCalled();
  });

  test('should return null on change if onClear prop is not passed', () => {
    const result = Toast.defaultProps?.onClear && Toast.defaultProps.onClear();
    expect(result).toBeNull();
  });
});
