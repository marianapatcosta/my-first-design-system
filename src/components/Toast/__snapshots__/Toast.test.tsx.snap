// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders without errors and matches snapshot 1`] = `
<div
  class="sc-hKgILt cRTYxH"
  id=""
  type="info"
>
  <button
    class="sc-bdfBwQ kNApjs sc-eCssSg hrInoB"
    id=""
    type="button"
  >
    <img
      alt=""
      class="sc-dlfnbm kQHvsZ"
      data-testid="button-icon"
      label=""
      src="close.svg"
    />
    
  </button>
  <img
    alt="info.svg icon"
    class="sc-jSgupP kbWecE"
    src="info.svg"
  />
  <div
    class="sc-gKsewC eEKOfY"
    type="info"
  >
    info toast
  </div>
</div>
`;
